src/middlewared/middlewared/plugins/pool.py

```diff
@@ -1077,6 +1079,124 @@ def __find_disk_from_topology(self, label, pool):
+    async def unlock_services_restart_choices(self):
+        svcs = {
+            'afp': 'AFP',
+            'cifs': 'SMB',
+            'ftp': 'FTP',
+            'iscsitarget': 'iSCSI',
+            'nfs': 'NFS',
+            'webdav': 'WebDAV',
+            'jails': 'Jails/Plugins',
+        }
+        return svcs
+
+    @item_method
+    @accepts(Int('id'), Dict(
+        'options',
+        Str('passphrase', password=True, required=False),
+        Bool('recoverykey', default=False),
+        List('services_restart', default=[]),
+    ))
+    @job(lock='unlock_pool', pipes=['input'], check_pipes=False)
+    async def unlock(self, job, oid, options):
+        """
+        Unlock encrypted pool `id`.
+        """
+        pool = await self._get_instance(oid)
+
+        verrors = ValidationErrors()
+
+        if pool['encrypt'] == 0:
+            verrors.add('id', 'Pool is not encrypted.')
+        elif pool['status'] != 'OFFLINE':
+            verrors.add('id', 'Pool already unlocked.')
+
+        if options.get('passphrase') and options['recoverykey']:
+            verrors.add(
+                'options.passphrase', 'Either provide a passphrase or a recovery key, not both.'
+            )
+        elif not options.get('passphrase') and not options['recoverykey']:
+            verrors.add(
+                'options.passphrase', 'Provide a passphrase or a recovery key.'
+            )
+
+        services_restart_choices = set((await self.unlock_services_restart_choices()).keys())
+        options_services_restart = set(options['services_restart'])
+        invalid_choices = options_services_restart - services_restart_choices
+        if invalid_choices:
+            verrors.add(
+                'options.services_restart', f'Invalid choices: {", ".join(invalid_choices)}'
+            )
+
+        if verrors:
+            raise verrors
+
+        if options['recoverykey']:
+            job.check_pipe("input")
+            with tempfile.NamedTemporaryFile(mode='wb+', dir='/tmp/') as f:
+                f.write(job.pipes.input.r.read())
+                f.flush()
+                failed = await self.middleware.call('disk.geli_attach', pool, None, f.name)
+        else:
+            failed = await self.middleware.call('disk.geli_attach', pool, options['passphrase'])
+
+        # We need to try to import the pool even if some disks failed to attach
+        try:
+            await self.middleware.call('zfs.pool.import_pool', pool['guid'], {
+                'altroot': '/mnt',
+                'cachefile': ZPOOL_CACHE_FILE,
+            })
+        except ZFSException as e:
+            # mounting filesystems may fail if we have readonly datasets as parent
+            if e.code.name != 'MOUNTFAILED':
+                raise
+        except Exception as e:
+            if failed > 0:
+                raise CallError(
+                    f'Pool could not be imported: {failed} devices failed to decrypt.'
+                )
+            raise
```

Comment:
Don't we need to rollback `geli.attach` here?

Reply:
Not sure, this is a direct move of code, not reimplementation.

I could see advantages of detaching the providers but also disadvantage on debugging in case of problems. Can you think of a bigger downside in not doing so?

Reply:
I know nothing about how geli is used in real life and what could cause issues when using it but it feels like if something that requires key was left unlocked at least partially, user should have explicit and clear feedback on that. So if we could unlock some devices and had left them in that state, we should warn the user. Just extending the message with warning will probably be fine.

Reply:
@themylogin fixed